### ISSUE TITLE
Improve Frontend Support for Errors

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#5026]
This is a tracking issue to improve how the frontend copes with errors.

The goal is that compiler should be able to progress through each phase, even if errors are encountered. 

Actionable steps (to be completed in separate PRs):

- [x] Refactor `Validation.Success` to have the shape `case class Success[T, E](t: T, errors: LazyList[E])`.
- [x] Change `Main.scala`, `Shell.scala`, `Lsp.scala`, `TestFlix.scala` (etc) to treat `Success(_, someNonEmptyList)` as failure. In other words, `Success` is only success if there are no errors in the lazy list.
- [x] Introduce `Validation.toSoftFailure(T, E)` which creates a `Success` with an element `T` and an error `E`.
- [x] Modify `LanguageServer` to update root AST even in the presence of errors.
- [x] Introduce a new `case class Expression.Error(m: CompilationMessage)` into all frontend ASTs (`WeededAst` until `LoweredAst`). This should be last expression in every AST.
- [x] Incrementally modify all frontend phases from `Weeder` to just before `Lowering` to return an `Error` expression and use `toSoftFailure` instead of `toFailure`. I suggest to this in many many smaller PRs (each at most a few hundred lines). Begin with just a single error, e.g. `IllegalInt` to ensure that the pipeline works.
- [ ] Rewrite the following phases to use `Validation`:
  - [ ] `Regions`
  - [x] `PatternExhaustiveness`
  - [x] `Redundancy`
  - [ ] `Safety`

**Comment by JonathanStarup** [comment]
Will this allow vscode to use the parsed AST even though the types fails for example?

**Comment by magnus-madsen** [comment]
> Will this allow vscode to use the parsed AST even though the types fails for example?

It will allow most of VSCode code work even if there are some non-parsing related errors.

**Comment by JonathanStarup** [comment]
Having a stupid/unsound robust parser as the final fallback could also be interesting. That would allow some highlighting and possibly better autocomplete on partially broken programs.

**Comment by magnus-madsen** [comment]
> Having a stupid/unsound robust parser as the final fallback could also be interesting. That would allow some highlighting and possibly better autocomplete on partially broken programs.

There is a separate issue for that.

**Comment by jaschdoc** [comment]
Should we maybe introduce a `SoftError` case class? I know this will affect the entire compiler, but might be simpler (and safer) in the long run than having to check `errors.isEmpty` at call sites.

**Comment by mlutze** [comment]
> Should we maybe introduce a `SoftError` case class? I know this will affect the entire compiler, but might be simpler (and safer) in the long run than having to check `errors.isEmpty` at call sites.

I agree with this.

**Comment by magnus-madsen** [comment]
> Should we maybe introduce a `SoftError` case class? I know this will affect the entire compiler, but might be simpler (and safer) in the long run than having to check `errors.isEmpty` at call sites.

Let us try it and see how it looks.

**Comment by magnus-madsen** [comment]
We have to be very careful to fail whenever an error actually occurs.

**Comment by magnus-madsen** [comment]
(Note that several tasks in this can be worked in parallel, if desired :))

**Comment by magnus-madsen** [comment]
@jaschdoc I think as you go through this, you should:

- Do it by phase.
- Limit yourself to ~15-20 errors at a time. (i.e. per PR).
- Check that every error has a corresponding test case.
- Since you are doing the last, it might make sense to reorganize each `TestX` if something is out of order and not fully up to date.

**Comment by mlutze** [comment]
> @jaschdoc I think as you go through this, you should:
> 
>     * Do it phase by phase.
> 
>     * Limit yourself to ~15-20 errors at a time. (i.e. per PR).
> 
>     * Check that every error has a corresponding test case.
> 
>     * Since you are doing the last, it might make sense to reorganize each `TestX` if something is out of order and not fully up to date.

If we're renaming tests, could we also make the names more descriptive? Sometimes it's not clear what a test is really testing.

**Comment by magnus-madsen** [comment]
I would keep the names for now, i.e. the test name should match the error name.

But we should have tests for everything in that should occur in some consistent order.

**Comment by magnus-madsen** [comment]
@jaschdoc In the interest of handling more errors, I think we should add this function to Validation:

```
  /**
    * Transform exactly one hard error into a soft error using the given function `f`.
    */
  def softRecoverOne[U >: T](f: E => U): Validation[U, E] = this match {
    case Validation.Failure(errors) if errors.length == 1 =>
      val one = errors.head
      Validation.SoftFailure(f(one), LazyList(one))
    case _ => this
  }
```

after `toHardFailure`

Then you can change e.g. `Weeder` to:

```
    case ParsedAst.Expression.Lit(sp1, lit, sp2) =>
      mapN(weedLiteral(lit)) {
        case l => WeededAst.Expression.Cst(l, mkSL(sp1, sp2))
      }.softRecoverOne(m => WeededAst.Expression.Error(m))
```

which means that should the call to `weedLiteral` fail (with one hard error) then we recover with a soft error.

**Comment by jaschdoc** [comment]
Refactor phases to use soft failures
- [ ] Weeder
- [ ] Namer
- [ ] Resolver
- [ ] ...

Refactor following phases to use Validation
- [ ] Regions
- [ ] PatternExhaustiveness
- [ ] Redundancy
- [ ] Safety

**Comment by magnus-madsen** [comment]
@jaschdoc Did you see #5421?

I would like to get a move on this because I have a new student project that might depend on (some of) this infrastructure :)

**Comment by jaschdoc** [comment]
> @jaschdoc Did you see #5421?
> 
> I would like to get a move on this because I have a new student project that might depend on (some of) this infrastructure :)

Yeah, makes sense :)

**Comment by jaschdoc** [comment]
Do we still keep `softRecoverOne`? I think it may be very useful in some places unless we completely rewrite `Validation`. The problem with `recoverOne` and subtyping is that if a helper function returns a validation with some subtype of say `WeederError`. Then if it calls another helper function that returns a validation we cannot map that to validation of the desired subtype of `WeederError` (especially since it's impossible to recover from an error in a helper function). Does that make sense?

**Comment by jaschdoc** [comment]
What was the reason for introducing `recoverOne` instead of `softRecoverOne`?

**Comment by jaschdoc** [comment]
Of course we can just replace `softRecoverOne` uses with `recoverOne { case err: Error  => ??? }` for now if we want to switch to that immediately

**Comment by magnus-madsen** [comment]
> What was the reason for introducing recoverOne instead of softRecoverOne?

The reason is that `recoverOne` forces you to declare the error you can recover from. Its similar to how you are not supposed to use a generic try-catch, but only catch the exception you expect. 

I think we should migrate towards `recoverOne` unless it becomes to burdensome.

**Comment by magnus-madsen** [comment]
But indeed limitations to Scala's type inference makes this annoying.

**Comment by magnus-madsen** [comment]
Continued in https://github.com/flix/flix/issues/5501 but with a different strategy.
